File: FinalProject/1_IntegerPartition/Baitoan3/ConjAndOdd.cpp
#include <bits/stdc++.h>
using namespace std;

bool isSelfConjugate(const vector<int>& parts) {
    int m = (int)parts.size();
    vector<int> conjugate(m, 0);
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < m && parts[j] >= i + 1; ++j) {
            ++conjugate[i];
        }
    }
    for (int i = 0; i < m; ++i) {
        if (parts[i] != conjugate[i]) return false;
    }
    return true;
}

void genSelfConjugatePartitions(int n, int k, int maxVal, vector<int>& curr, vector<vector<int>>& result) {
    if (k == 0 && n == 0) {
        if (isSelfConjugate(curr)) {
            result.push_back(curr);
        }
        return;
    }
    if (k <= 0 || n <= 0) return;

    for (int i = min(maxVal, n); i >= 1; --i) {
        curr.push_back(i);
        genSelfConjugatePartitions(n - i, k - 1, i, curr, result);
        curr.pop_back();
    }
}

void genAllSelfConjugatePartitions(int n, int maxVal, vector<int>& curr, vector<vector<int>>& result) {
    if (n == 0) {
        if (isSelfConjugate(curr)) {
            result.push_back(curr);
        }
        return;
    }
    for (int i = min(maxVal, n); i >= 1; --i) {
        curr.push_back(i);
        genAllSelfConjugatePartitions(n - i, i, curr, result);
        curr.pop_back();
    }
}

int countOddParts(int n) {
    vector<int> dp(n + 1, 0);
    dp[0] = 1;
    for (int odd = 1; odd <= n; odd += 2) {
        for (int i = odd; i <= n; ++i) {
            dp[i] += dp[i - odd];
        }
    }
    return dp[n];
}

void genOddPartitions(int n, int minOdd, vector<int>& curr, vector<vector<int>>& result) {
    if (n == 0) {
        result.push_back(curr);
        return;
    }
    if (n < 0) return;

    for (int odd = minOdd; odd <= n; odd += 2) {
        curr.push_back(odd);
        genOddPartitions(n - odd, odd, curr, result);
        curr.pop_back();
    }
}

void genDistinctOddPartitions(int n, int start, vector<int>& curr, vector<vector<int>>& result) {
    if (n == 0) {
        result.push_back(curr);
        return;
    }
    if (n < 0) return;

    for (int odd = start; odd <= n; odd += 2) {
        curr.push_back(odd);
        genDistinctOddPartitions(n - odd, odd + 2, curr, result);
        curr.pop_back();
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // Tổng số phân hoạch tự liên hợp của n
    vector<vector<int>> allSelfConjParts;
    vector<int> curr;
    for (int partsCount = 1; partsCount <= n; ++partsCount) {
        genSelfConjugatePartitions(n, partsCount, n, curr, allSelfConjParts);
    }

    cout << "\n=== Tong so phan hoach tu lien hop cua " << n << ": " << allSelfConjParts.size() << "\n";
    for (const auto& part : allSelfConjParts) {
        for (int i = 0; i < (int)part.size(); ++i) {
            cout << part[i];
            if (i < (int)part.size() - 1) cout << " + ";
        }
        cout << "\n";
    }

    // Phân hoạch thành odd parts
    int oddPartsCount = countOddParts(n);
    cout << "\n=== So phan hoach cua " << n << " co phan le: " << oddPartsCount << "\n";
    vector<vector<int>> oddParts;
    curr.clear();
    genOddPartitions(n, 1, curr, oddParts);

    for (auto& part : oddParts) {
        sort(part.rbegin(), part.rend());
    }
    sort(oddParts.begin(), oddParts.end());

    for (const auto& part : oddParts) {
        for (int i = 0; i < (int)part.size(); ++i) {
            cout << part[i];
            if (i < (int)part.size() - 1) cout << " + ";
        }
        cout << "\n";
    }

    // Phân hoạch thành distinct odd parts
    vector<vector<int>> distinctOddParts;
    curr.clear();
    genDistinctOddPartitions(n, 1, curr, distinctOddParts);

    cout << "\n=== So phan hoach cua " << n << " thanh cac phan le phan biet: " << distinctOddParts.size() << "\n";

    for (auto& part : distinctOddParts) {
        sort(part.rbegin(), part.rend());
    }
    sort(distinctOddParts.begin(), distinctOddParts.end());

    for (const auto& part : distinctOddParts) {
        for (int i = 0; i < (int)part.size(); ++i) {
            cout << part[i];
            if (i < (int)part.size() - 1) cout << " + ";
        }
        cout << "\n";
    }

    // So sánh
    cout << "\n=== So sanh:\n";
    cout << "Tong so phan hoach tu lien hop = " << allSelfConjParts.size() << "\n";
    cout << "So phan hoach distinct odd parts = " << distinctOddParts.size() << "\n";
    if (allSelfConjParts.size() == distinctOddParts.size()) {
        cout << "=> Bang nhau";
    } else {
        cout << "=> Khac nhau\n";
    }

    return 0;
}

File: FinalProject/1_IntegerPartition/Baitoan3/ConjAndOdd.py
def is_self_conjugate(parts):
    m = len(parts)
    conjugate = [0] * m
    for i in range(m):
        for j in range(m):
            if parts[j] >= i + 1:
                conjugate[i] += 1
    return parts == conjugate

def gen_self_conjugate_partitions(n, k, max_val, curr, result):
    if k == 0 and n == 0:
        if is_self_conjugate(curr):
            result.append(curr[:])
        return
    if k <= 0 or n <= 0:
        return
    for i in range(min(max_val, n), 0, -1):
        curr.append(i)
        gen_self_conjugate_partitions(n - i, k - 1, i, curr, result)
        curr.pop()

def gen_odd_partitions(n, min_odd, curr, result):
    if n == 0:
        result.append(curr[:])
        return
    if n < 0:
        return
    for odd in range(min_odd, n + 1, 2):
        curr.append(odd)
        gen_odd_partitions(n - odd, odd, curr, result)
        curr.pop()

def count_odd_partitions(n):
    dp = [0] * (n + 1)
    dp[0] = 1
    for odd in range(1, n + 1, 2):
        for i in range(odd, n + 1):
            dp[i] += dp[i - odd]
    return dp[n]

def gen_distinct_odd_partitions(n, start, curr, result):
    if n == 0:
        result.append(curr[:])
        return
    if n < 0:
        return
    for odd in range(start, n + 1, 2):
        curr.append(odd)
        gen_distinct_odd_partitions(n - odd, odd + 2, curr, result)
        curr.pop()

n = int(input())

# Tự liên hợp
all_self_conj_parts = []
curr = []
for k in range(1, n + 1):
    gen_self_conjugate_partitions(n, k, n, curr, all_self_conj_parts)

print(f"\n=== Tổng số phân hoạch tự liên hợp của {n}: {len(all_self_conj_parts)}")
for part in all_self_conj_parts:
    print(" + ".join(map(str, part)))

# Phân hoạch toàn số lẻ
odd_parts = []
curr = []
gen_odd_partitions(n, 1, curr, odd_parts)
for part in odd_parts:
    part.sort(reverse=True)
odd_parts.sort()

print(f"\n=== Số phân hoạch của {n} có phần tử lẻ: {len(odd_parts)}")
for part in odd_parts:
    print(" + ".join(map(str, part)))

# Phân hoạch lẻ phân biệt
distinct_odd_parts = []
curr = []
gen_distinct_odd_partitions(n, 1, curr, distinct_odd_parts)
for part in distinct_odd_parts:
    part.sort(reverse=True)
distinct_odd_parts.sort()

print(f"\n=== Số phân hoạch của {n} thành các phần lẻ phân biệt: {len(distinct_odd_parts)}")
for part in distinct_odd_parts:
    print(" + ".join(map(str, part)))

# So sánh
print("\n=== So sánh:")
print(f"Tổng số phân hoạch tự liên hợp = {len(all_self_conj_parts)}")
print(f"Số phân hoạch lẻ phân biệt = {len(distinct_odd_parts)}")
if len(all_self_conj_parts) == len(distinct_odd_parts):
    print("=> Bằng nhau")
else:
    print("=> Khác nhau")